File: apps/worker/src/taint.ts
import type { ObservationInput, TaintClass } from "@argus/shared";

/**
 * Lightweight taint inference for storage/UI (mirrors the detection service's
 * taint.py). The detection service remains the source of truth for security
 * decisions; this just labels stored observations so the trace view can tint
 * untrusted spans without a round-trip.
 */
export function inferTaint(
  obs: ObservationInput,
  toolOverrides: Record<string, string> = {},
): TaintClass {
  if (obs.taint) return obs.taint;
  if (obs.type === "retrieval" || obs.type === "tool") {
    const key = obs.taintSource || obs.name;
    if (toolOverrides[key] === "trusted" || toolOverrides[obs.name] === "trusted")
      return "system";
    return "untrusted_external";
  }
  const role = (obs.role || "").toLowerCase();
  if (role === "system") return "system";
  if (role === "user") return "user";
  if (role === "assistant" || role === "model") return "model";
  if (obs.type === "generation") return "model";
  return "user";
}

File: services/detection/argus_detection/taint.py
"""Taint classification (docs/04 §Trust model).

Assigns a trust class to each observation when the SDK didn't label it, applying
operator tool-overrides. Retrieval outputs and tool results default to
`untrusted_external`; everything else falls back to its role/type.
"""
from __future__ import annotations

from .models import Observation, ObservationType, TaintClass


def classify(obs: Observation, tool_overrides: dict[str, str] | None = None) -> TaintClass:
    # Explicit SDK label always wins.
    if obs.taint is not None:
        return obs.taint

    overrides = tool_overrides or {}

    # Tool/retrieval spans: untrusted by default, unless operator overrode this
    # specific tool/source to 'trusted'.
    if obs.type in (ObservationType.retrieval, ObservationType.tool):
        key = obs.taint_source or obs.name
        override = overrides.get(key) or overrides.get(obs.name)
        if override == "trusted":
            return TaintClass.system  # treated as trusted content
        return TaintClass.untrusted_external

    # Role-based for chat spans.
    role = (obs.role or "").lower()
    if role == "system":
        return TaintClass.system
    if role == "user":
        return TaintClass.user
    if role in ("assistant", "model"):
        return TaintClass.model

    if obs.type == ObservationType.generation:
        return TaintClass.model

    # Unknown → treat as user input (the conservative default for scanning).
    return TaintClass.user
